src/sentry/hybridcloud/models/outbox.py

```diff
@@ -219,11 +220,143 @@ def process_shard(self, latest_shard_row: OutboxBase | None) -> Generator[Outbox
 
             yield next_shard_row
 
+    @contextlib.contextmanager
+    def reserve_coalesced_messages(
+        self,
+        latest_shard_row: OutboxBase | None = None,
+        is_synchronous_flush: bool = False,
+    ) -> Generator[OutboxBase | None]:
+        """
+        Reserve a slice of coalesced outboxes based on latest_shard_row
+
+        1. We need to reserve the group of messages by getting locks on the top
+        and bottom of the coalesce group.
+        2. We advance the schedule_for of all messages in the coalesce group. This
+        combined with the locks from 1. prevent concurrent deliveries of the same messages.
+        3. We commit the update transaction, and call the signal handler. Signal handler
+        need to be called outside of transactions as they frequently use RPC which
+        can have long latency. If a transaction was held during this time it could
+        hit statement timeouts and the transaction would be aborted.
+        4. Once the signal handler completes, we delete the coalesced messages
+        as they have been 'delivered'.
+
+        If no messages could be reserved or there was contention, this will return None.
+        """
+        using = router.db_for_write(type(self))
+        tags: dict[str, int | str] = {"category": "None", "synchronous": int(is_synchronous_flush)}
+
+        # If we have a latest_row, only operate on messages that were created before it.
+        filters = {} if latest_shard_row is None else dict(id__lte=latest_shard_row.id)
+        try:
+            now = timezone.now()
+            with transaction.atomic(using=using, savepoint=False):
+                # Lock all the rows we're going to process and attempt to reserve them.
+                reserved_ids = list(
+                    self.select_coalesced_messages()
+                    .filter(scheduled_for__lte=now, **filters)
+                    .select_for_update(nowait=True)
+                    .order_by("id")
+                    .values_list("id", flat=True)
+                )
```

Comment:
There is still one locking query here. We need to lock all the rows that are part of the coalesce group to prevent concurrent writes.